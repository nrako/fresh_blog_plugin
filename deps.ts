export {
  DOMParser,
} from 'https://deno.land/x/deno_dom@v0.1.43/deno-dom-wasm.ts'
export { parse } from 'https://deno.land/x/xml@2.1.3/mod.ts'
export { type ParserOptions } from 'https://deno.land/x/xml@2.1.3/utils/types.ts'
