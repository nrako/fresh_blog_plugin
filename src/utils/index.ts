import { type InternalOptions } from '../../mod.ts'

export function getFeedPathPrefix(
  blogOptions: InternalOptions,
): string {
  return blogOptions.feedPathPrefix ?? blogOptions.path
}
