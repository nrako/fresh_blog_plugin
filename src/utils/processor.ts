import { transformerNotationDiff } from 'npm:shikiji-transformers@0.9'
import { VFile } from 'https://esm.sh/vfile@5'
import { mystParse } from 'https://esm.sh/myst-parser@1.0.22'
import {
  basicTransformationsPlugin,
  DOITransformer,
  getFrontmatter,
  GithubTransformer,
  inlineMathSimplificationPlugin,
  linksPlugin,
  RRIDTransformer,
  WikiTransformer,
} from 'https://esm.sh/myst-transforms@1.2.1'
import { unified } from 'https://esm.sh/unified@10'
import { visit } from 'https://esm.sh/unist-util-visit@5'
import { mystToHtml } from 'https://esm.sh/myst-to-html@1.0.22'
import rehypeShiki from 'https://esm.sh/@shikijs/rehype@1.0.0-beta.3'
import rehypeStringify from 'https://esm.sh/v135/rehype-stringify@8'
import rehypeParse from 'https://esm.sh/rehype-parse@8'
import rehypeExternalLinks from 'https://esm.sh/rehype-external-links@3'
import rehypeKatex from 'https://esm.sh/v135/rehype-katex@7.0.0/lib/index.js'
import rehypeAutolinkHeadings from 'https://esm.sh/rehype-autolink-headings@7'
import { h } from 'https://esm.sh/hastscript@9'
import { BlogOptions, defaultOptions } from '../../mod.ts'
import { validatePageFrontmatter } from 'https://esm.sh/myst-frontmatter@1.1.23'

export type ParseOptions = Required<Pick<BlogOptions, 'highlighter'>>

interface Message {
  property: string
  message: string
}

export interface Messages {
  errors?: Message[]
  warnings?: Message[]
}

async function parse(text: string, options: ParseOptions) {
  const file = new VFile()
  const mdast = mystParse(text, {
    markdownit: { linkify: true },
    vfile: file,
  })

  const messages: Messages = {}
  const vfile = new VFile()
  const { frontmatter: rawPageFrontmatter } = getFrontmatter(vfile, mdast, {
    propagateTargets: true,
  })
  const frontmatter = validatePageFrontmatter(
    rawPageFrontmatter,
    { property: 'frontmatter', messages },
  )

  const linkTransforms = [
    new WikiTransformer(),
    new GithubTransformer(),
    new DOITransformer(),
    new RRIDTransformer(),
  ]

  // For the mdast that we show, duplicate, strip positions and dump to yaml
  // Also run some of the transforms, like the links
  const mdastPre = JSON.parse(JSON.stringify(mdast))
  unified()
    .use(basicTransformationsPlugin, {})
    .use(inlineMathSimplificationPlugin)
    .use(linksPlugin, { transformers: linkTransforms })
    .runSync(mdastPre)

  visit(mdastPre, (n) => delete n.position)
  const htmlString = mystToHtml(mdastPre, {
    hast: {
      allowDangerousHtml: true,
    },
    stringifyHtml: {
      allowDangerousHtml: true,
    },
  })

  const r = await unified()
    .use(rehypeParse, { fragment: true })
    .use(rehypeShiki, {
      ...options.highlighter,
      transformers: [
        transformerNotationDiff(),
      ],
    })
    // @ts-ignore required until unified can be upgrade to v11 when mystmd plays fair
    .use(rehypeExternalLinks, { rel: ['noopener'] })
    // @ts-ignore required until unified can be upgrade to v11 when mystmd plays fair
    .use(rehypeKatex)
    .use(rehypeAutolinkHeadings, {
      content() {
        return [
          h('span.anchorsign', { ariaHidden: 'true' }),
        ]
      },
    })
    // @ts-ignore required until unified can be upgrade to v11 when mystmd plays fair
    .use(rehypeStringify)
    .process(htmlString)

  const html = r.value

  return { frontmatter, html, messages }
}

export default async function processor(
  content: string,
  options: ParseOptions = defaultOptions,
) {
  return await parse(content, options)
}
