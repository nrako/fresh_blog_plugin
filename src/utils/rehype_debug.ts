import type { Node } from 'https://esm.sh/v135/@types/unist@3.0.2/index.d.ts'
import type { Element } from 'npm:@types/hast'
import { visit } from 'https://esm.sh/unist-util-visit@5'

export default () => {
  function visitor(
    node: Element,
    _index: number,
    parent: Element | null,
  ) {
    console.log(node)
    if (node.tagName !== 'code') return
    if (!parent || parent.tagName !== 'pre') return
  }

  return (tree: Node) => {
    visit(tree, 'element', visitor)
  }
}
