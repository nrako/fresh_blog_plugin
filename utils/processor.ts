import { render } from 'gfm/mod.ts'

export default async function processor(content: string) {
  await import(
    'https://esm.sh/prismjs@1.29.0/components/prism-typescript?no-check'
  )

  const html = render(content, {
    allowMath: true,
    allowIframes: true,

    // TODO use front matter post.attrs.disableHTMLSanitzation or
    // post.attrs.enableTwitterEmbed? Might not be sufficient since
    // gfm also sanitize CSS classes which could be useful for
    // YouTube iframes...

    // NOTE this is actually mostly needed because
    // https://deno.land/x/gfm@0.3.0/mod.ts?source is quite
    // opiniated about which HTML elements and CSS classes can make
    // it through. For YouTube `w-full aspect-video" CSS classes would not go through.
    // For Twitter `twitter-tweet` on `<blockquote>` CSS class would not go through
    disableHtmlSanitization: true,
  })

  return html
}
