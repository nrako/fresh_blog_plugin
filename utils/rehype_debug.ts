import { visit } from 'https://esm.sh/unist-util-visit@5'
import { Node } from 'https://esm.sh/v135/@types/unist@3.0.2/index.d.ts'
import { Element } from 'npm:@types/hast'

export default () => {
  function visitor(
    node: Element,
    _index: number,
    parent: Element | null,
  ) {
    console.log(node)
    if (node.tagName !== 'code') return
    if (!parent || parent.tagName !== 'pre') return
  }

  return (tree: Node) => {
    visit(tree, 'element', visitor)
  }
}
