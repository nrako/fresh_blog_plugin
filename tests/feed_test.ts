import { assertSnapshot } from '$fresh/src/server/deps.ts'
import { type BlogOptions } from '../mod.ts'
import { createFreshBlogHandler } from './utils.ts'
import { parse, type ParserOptions } from '../deps.ts'

const blogOptions: BlogOptions = {
  contentDir: './tests/fixture/posts/',
}

const parserOptions: ParserOptions = {
  reviver({ value, key: _key, tag, properties: _properties }) {
    // replace the year in the copyright with placerholder
    if (tag === 'rights' && typeof value === 'string') {
      return value.replace(/20\d{2}/, 'YYYY')
    }
    return value
  },
}

Deno.test('Atom 1.0 feed match snapshot', async (t) => {
  const handler = await createFreshBlogHandler(blogOptions)
  const resp = await handler(
    new Request('http://127.0.0.1/blog/atom'),
  )
  const body = parse(await resp.text(), parserOptions)
  await assertSnapshot(t, body)
})

Deno.test('JSON Feed 1.0 feed match snapshot', async (t) => {
  const handler = await createFreshBlogHandler(blogOptions)
  const resp = await handler(
    new Request('http://127.0.0.1/blog/json'),
  )
  const body = await resp.text()
  await assertSnapshot(t, body)
})

Deno.test('RSS 2.0 feed match snapshot', async (t) => {
  const handler = await createFreshBlogHandler(blogOptions)
  const resp = await handler(
    new Request('http://127.0.0.1/blog/rss'),
  )
  const body = parse(await resp.text(), parserOptions)
  await assertSnapshot(t, body)
})
