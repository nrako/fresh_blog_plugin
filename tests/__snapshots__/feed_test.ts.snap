export const snapshot = {};

snapshot[`Atom 1.0 feed match snapshot 1`] = `
{
  feed: {
    "@xmlns": "http://www.w3.org/2005/Atom",
    entry: [
      {
        id: "http://127.0.0.1/Post X",
        link: {
          "#text": null,
          "@href": "http://127.0.0.1/blog/post_x",
        },
        published: "2024-01-06T22:38:45.000Z",
        rights: "Copyright YYYY http://127.0.0.1",
        summary: {
          "#text": "desc",
          "@type": "html",
        },
        title: {
          "#text": "Post X",
          "@type": "html",
        },
        updated: "2024-01-06T22:38:45.000Z",
      },
      {
        id: "http://127.0.0.1/Hello World",
        link: {
          "#text": null,
          "@href": "http://127.0.0.1/blog/hello_world",
        },
        published: "2024-01-02T22:38:45.000Z",
        rights: "Copyright YYYY http://127.0.0.1",
        summary: {
          "#text": "yello, description",
          "@type": "html",
        },
        title: {
          "#text": "Hello World",
          "@type": "html",
        },
        updated: "2024-01-02T22:38:45.000Z",
      },
    ],
    generator: "Feed (https://github.com/jpmonette/feed) for Deno",
    icon: "http://127.0.0.1/favicon.ico",
    id: "http://127.0.0.1/blog",
    link: [
      {
        "#text": null,
        "@href": "http://127.0.0.1/blog",
        "@rel": "alternate",
      },
      {
        "#text": null,
        "@href": "http://127.0.0.1/blog/atom",
        "@rel": "self",
      },
    ],
    rights: "Copyright YYYY http://127.0.0.1",
    subtitle: "This is a Fresh Blog",
    title: "Blog",
    updated: "2024-01-06T22:38:45.000Z",
  },
  xml: {
    "@encoding": "utf-8",
    "@version": 1,
  },
}
`;

snapshot[`JSON Feed 1.0 feed match snapshot 1`] = `
'{
    "version": "https://jsonfeed.org/version/1",
    "title": "Blog",
    "home_page_url": "http://127.0.0.1/blog",
    "feed_url": "http://127.0.0.1/blog/json",
    "description": "This is a Fresh Blog",
    "items": [
        {
            "id": "http://127.0.0.1/Post X",
            "url": "http://127.0.0.1/blog/post_x",
            "title": "Post X",
            "summary": "desc",
            "date_modified": "2024-01-06T22:38:45.000Z",
            "date_published": "2024-01-06T22:38:45.000Z"
        },
        {
            "id": "http://127.0.0.1/Hello World",
            "url": "http://127.0.0.1/blog/hello_world",
            "title": "Hello World",
            "summary": "yello, description",
            "date_modified": "2024-01-02T22:38:45.000Z",
            "date_published": "2024-01-02T22:38:45.000Z"
        }
    ]
}'
`;

snapshot[`RSS 2.0 feed match snapshot 1`] = `
{
  rss: {
    "@version": 2,
    "@xmlns:atom": "http://www.w3.org/2005/Atom",
    channel: {
      "atom:link": {
        "#text": null,
        "@href": "http://127.0.0.1/blog/rss",
        "@rel": "self",
        "@type": "application/rss+xml",
      },
      copyright: "Copyright 2024 http://127.0.0.1",
      description: "This is a Fresh Blog",
      docs: "https://validator.w3.org/feed/docs/rss2.html",
      generator: "Feed (https://github.com/jpmonette/feed) for Deno",
      item: [
        {
          description: "desc",
          guid: "http://127.0.0.1/Post X",
          link: "http://127.0.0.1/blog/post_x",
          pubDate: "Sat, 06 Jan 2024 22:38:45 GMT",
          title: "Post X",
        },
        {
          description: "yello, description",
          guid: "http://127.0.0.1/Hello World",
          link: "http://127.0.0.1/blog/hello_world",
          pubDate: "Tue, 02 Jan 2024 22:38:45 GMT",
          title: "Hello World",
        },
      ],
      language: "en",
      lastBuildDate: "Sat, 06 Jan 2024 22:38:45 GMT",
      link: "http://127.0.0.1/blog",
      title: "Blog",
    },
  },
  xml: {
    "@encoding": "utf-8",
    "@version": 1,
  },
}
`;
